File: src/main/groovy/net/zomis/brainf/model/ast/tree/ChangePointerSyntax.java
package net.zomis.brainf.model.ast.tree;

import net.zomis.brainf.model.BrainfuckRunner;

public class ChangePointerSyntax extends Syntax implements SteppableSyntax {

    private final int value;

    public ChangePointerSyntax(int value) {
        this.value = value;
    }

    @Override
    public int getValue() {
        return value;
    }

    @Override
    public void perform(BrainfuckRunner runner) {
        performTimes(runner, Math.abs(value));
    }

    @Override
    public String toString() {
        return "ChangePointerSyntax{" +
                "value=" + value +
                '}';
    }

    @Override
    public void performTimes(BrainfuckRunner runner, int steps) {
        steps = value >= 0 ? steps : -steps;
        runner.getMemory().setMemoryIndex(runner.getMemory().getMemoryIndex() + steps);
    }

}

File: brainduck-kotlin/src/commonMain/kotlin/net/zomis/brainf/model/ast/tree/ChangePointerSyntax.kt
package net.zomis.brainf.model.ast.tree;

import net.zomis.brainf.model.BrainfuckRunner;
import kotlin.math.absoluteValue

class ChangePointerSyntax(val value: Int) : Syntax(), SteppableSyntax {

    override fun getValue(): Int = value

    override fun perform(runner: BrainfuckRunner) {
        performTimes(runner, value.absoluteValue)
    }

    override fun toString(): String {
        return "ChangePointerSyntax{" +
                "value=" + value +
                '}';
    }

    override fun performTimes(runner: BrainfuckRunner, steps: Int) {
        val steps2 = if (value >= 0) steps else -steps
        runner.memory.setMemoryIndex(runner.memory.memoryIndex + steps2)
    }

}
